src/landscape.cpp

```diff
@@ -538,6 +539,8 @@ void DoClearSquare(TileIndex tile)
 	MakeClear(tile, CLEAR_GRASS, _generating_world ? 3 : 0);
 	MarkTileDirtyByTile(tile);
 	if (remove) RemoveDockingTile(tile);
+
+	InvalidateWaterRegion(tile);
```

Comment:
Looks like you now rely on DoClearSquare() being called when any tile is modified to refresh regions (except for objects). Is this always the case? Industries come to mind.

Reply:
The reason I use DoClearSquare() is that everything that wants to build something on a tile first has to check if that tile can be cleared before it can "claim it for itself". Industry construction does this too. The only exception I've found so far is placing objects, where tile clearing can't be done. I forgot the exact reason but it's in the comments in the place object command, something to do with NewGRFs.